packages/nextjs/components/assets/GithubIcon.tsx

```diff
@@ -0,0 +1,31 @@
+"use client";
+
+import { useEffect, useState } from "react";
+import { useTheme } from "next-themes";
+
+export const GithubIcon = ({ className }: { className: string }) => {
+  const { resolvedTheme } = useTheme();
+  const [isDarkMode, setIsDarkMode] = useState(false);
+
+  useEffect(() => {
+    if (resolvedTheme === "dark") {
+      setIsDarkMode(true);
+      return;
+    }
+
+    setIsDarkMode(false);
+  }, [resolvedTheme]);
```

Comment:
You don't need the useEffect, you can just do

```
  const { resolvedTheme } = useTheme();
  const isDarkMode = resolvedTheme === "dark";
```

Keep in mind that this will re-render when the mode changes (because we are using the useTheme hook)

Reply:
I liked the tip, but using it like this here, the initial render (f5) the resolvedTheme will be undefined. So it will set to false as default and only update the state when clicking on the button.

Do you have a way to avoid this? (I'll implement the dark: above, but tough it would be a nice technical conversation to have!)

Reply:
Damn, I think NextJS is giving us some shit haha.

Check this out
```tsx
"use client";

import { useTheme } from "next-themes";

export const GithubIcon = ({ className }: { className: string }) => {
  const { resolvedTheme } = useTheme();
  const isDarkMode = resolvedTheme === "dark";

  return (
    <>
      <p>{isDarkMode ? "fill-white" : "fill-black"}</p>
      <svg
        className={`${className} ${isDarkMode ? "fill-white" : "fill-black"}`}
        xmlns="http://www.w3.org/2000/svg"
        viewBox="0 0 16 16"
        width="16"
        height="16"
        fill="fff"
      >
        <path d="M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z"></path>
      </svg>
    </>
  );
};
```

This "works" (on F5 refresh) but throws a hydration error. 

![image](https://github.com/BuidlGuidl/batch4.buidlguidl.com/assets/2486142/48ba1b80-f9f5-4930-ae4b-2b945d324a38)

Probably related to SSR / Client side. I'm sure we'd have to do any obscure thing (isMounted? XD)

Tagging @technophile-04 here which I'm sure will throw some light (no pun intended) haha

Good discussion!

Reply:
The funny thing here is that if I remove the `<p>{isDarkMode ? "fill-white" : "fill-black"}</p>` that you created. It won't work on the first render. 

I wonder if next threat the arguments passing to a component differently than the content inside of it. As I remember it forces the page to re-render because the value of a content (<p>inside of a component</p>) changes. 

I always tough that It would behave the same for props passed to the component. It doesn't seems like it...

Reply:
> but using it like this here, the initial render (f5) the resolvedTheme will be undefined. 

This shouldn't be the case and `resolvedTheme` should always be present because the `next-theme` ThemeProvider automatically injects a script into next/head to update the html element with the correct attributes before rest of the page loads. Since our ThemeProvider is very at top of the component tree any child will always get `resolvedTheme` correct value on their first render itself. 


> but throws a hydration error.
> Probably related to SSR / Client side. I'm sure we'd have to do any obscure thing (isMounted? XD)

Yeah 💯😅, Even `next-theme` has this warning on their [README](https://github.com/pacocoursey/next-themes?tab=readme-ov-file#usetheme) : 
![Screenshot 2024-04-12 at 10 42 46 PM](https://github.com/BuidlGuidl/batch4.buidlguidl.com/assets/80153681/0f18ed57-da50-4342-bc6b-b553a85bbf09)

And [they suggest](https://github.com/pacocoursey/next-themes?tab=readme-ov-file#avoid-hydration-mismatch) same workaround as Carlos suggested using `isMounted`. 

People rely on cookies for handling theme to avoid this behavior

Reply:
Thanks for the explanations @technophile-04 ! I've investigated again and in fact it doesn't return undefined. 

Anyway we can see the same behavior in the [/builders page](https://batch4.buidlguidl.com/builders) . At least here if I have the dark theme selected and refresh, It will show the light theme. Even tried a different browser than brave.

I will all be solved when updating to the className='dark:xxxx' (the solution is already in the pr). But I'm still intrigued on why its behaving like this.

Reply:
> Anyway we can see the same behavior in the [/builders page](https://batch4.buidlguidl.com/builders) . At least here if I have the dark theme selected and refresh, It will show the light theme. Even tried a different browser than brave.

I see the dark mode but with a flash of gray color : 

And it expected behaviour since we are not using `isMounted` logic with `useTheme` hook there is hydration mismatch.Server is sending bg-gray class and hence initially it shows gray background and then turns into dark

Because when using `useTheme` we should delay rendering as mentioned here : https://github.com/BuidlGuidl/batch4.buidlguidl.com/pull/29#discussion_r1562920826